src/csharp/Grpc.Tools/DepFileUtil.cs

```diff
@@ -0,0 +1,198 @@
+C:\projects\foo\src\./foo.grpc.pb.h \
+C:\projects\foo\src\./foo.pb.cc \
+C:\projects\foo\src\./foo.pb.h: C:/foo/include/google/protobuf/wrappers.proto\
+ C:/foo/include/google/protobuf/any.proto\
+ C:/foo/include/google/protobuf/source_context.proto\
+ C:/foo/include/google/protobuf/type.proto\
+ foo.proto
+*/
+
+    // Read file names from the dependency file to the right of ':'.
+    public static string[] ReadDependencyInputs(string protoDepDir, string proto,
+                                                TaskLoggingHelper log) {
+      string depFilename = GetDepFilenameForProto(protoDepDir, proto);
+      string[] lines = ReadDepFileLines(depFilename, false, log);
+      if (lines.Length == 0) {
+        return lines;
+      }
+
+      var result = new List<string>();
+      bool skip = true;
+      foreach (string line in lines) {
+        // Start at the only line separating dependency outputs from inputs.
+        int ix = skip ? FindLineSeparator(line) : -1;
+        skip = skip && ix < 0;
+        if (skip) continue;
+        string file = ExtractFilenameFromLine(line, ix + 1, line.Length);
+        if (file == "") {
+          log.LogMessage(MessageImportance.Low,
+    $"Skipping unparsable dependency file {depFilename}.\nLine with error: '{line}'");
+          return new string[0];
+        }
+
+        // Do not bend over backwards trying not to include a proto into its
+        // own list of dependencies. Since a file is not older than self,
+        // it is safe to add; this is purely a memory optimization.
+        if (file != proto) {
+          result.Add(file);
+        }
+      }
+      return result.ToArray();
+    }
+
+    // Read file names from the dependency file to the left of ':'.
+    public static string[] ReadDependencyOutputs(string depFilename,
+                                                TaskLoggingHelper log) {
+      string[] lines = ReadDepFileLines(depFilename, true, log);
+      if (lines.Length == 0) {
+        return lines;
+      }
+
+      var result = new List<string>();
+      foreach (string line in lines) {
+        int ix = FindLineSeparator(line);
+        string file = ExtractFilenameFromLine(line, 0, ix >= 0 ? ix : line.Length);
+        if (file == "") {
+          log.LogError("Unable to parse generated dependency file {0}.\n" +
+                       "Line with error: '{1}'", depFilename, line);
+          return new string[0];
+        }
+        result.Add(file);
+
+        // If this is the line with the separator, do not read further.
+        if (ix >= 0)
+          break;
+      }
+      return result.ToArray();
+    }
+
+    // Get complete dependency file name from directory hash and file name,
+    // tucked onto protoDepDir, e. g.
+    // ("out", "foo/file.proto") => "out/deadbeef12345678_file.protodep".
+    // This way, the filenames are unique but still possible to make sense of.
+    public static string GetDepFilenameForProto(string protoDepDir, string proto) {
```

Comment:
not sure I understand the need for hashing the directory name here, it seems pretty artificial.

Reply:
I am adding this commentary, and I think it makes sense to reproduce here as I tried to explain it as well as I could:
> Since a project may contain proto files with the same filename but in different
> directories, a unique filename for the dependency file is constructed based on the
> proto file name both name and directory. The directory path can be arbitrary,
> for example, it can be outside of the project, or an absolute path including
> a drive letter, or a UNC network path. A name constructed from such a path by,
> for example, replacing disallowed name characters with an underscore, may well
> be over filesystem's allowed path length, since it will be located under the
> project and solution directories, which are also some level deep from the root.
> Instead of creating long and unwieldy names for these proto sources, we cache
> the full path of the name without the filename, and append the filename to it,
> as in e. g. "foo/file.proto" will yield the name "deadbeef12345678_file", where
> "deadbeef12345678" is a presumed hash value of the string "foo/". This allows
> the file names be short, unique (up to a hash collision), and still allowing
> the user to guess their provenance.

In other words, I need to encode distinct dependency file names of `foo.proto`, `sub1/sub2/foo.proto`, `../../../../protofiles/helloworld/hello.proto` and `\\server.company.local\fs\shared\packages\protofiles\helloword\hello.proto`. I do not want to end up with names like `__server.company.local_fs_shared_packages_protofiles_helloword_hello.protodep` because they may be arbitrarily long and just would not fit into filesystem's path length restriction when placed under `C:\users\firstname.lastname.OURDOMAIN\Documents\Visual Studio 2017\myprojects\SmartAction.SomethingLong.SomethingEvenLonger\SmartAction.SomethingLong.SomethingEvenLonger.Client\obj`. This is quite common for the last 3 path components to be like this, and may be not so rare for the rest of this not exactly very contrived path name; I have had real problems exactly like this (and my project directory is `c:\projects`, not the least *because* of this :) ). Also, they are ugly and no less clash-prone than 64-bit hashes.